frame/staking/src/lib.rs

```diff
@@ -719,23 +717,23 @@ decl_storage! {
 		pub Invulnerables get(fn invulnerables) config(): Vec<T::AccountId>;
 
 		/// Map from all locked "stash" accounts to the controller account.
-		pub Bonded get(fn bonded): map hasher(blake2_256) T::AccountId => Option<T::AccountId>;
+		pub Bonded get(fn bonded): map hasher(twox_64_concat) T::AccountId => Option<T::AccountId>;
 
 		/// Map from all (unlocked) "controller" accounts to the info regarding the staking.
 		pub Ledger get(fn ledger):
-			map hasher(blake2_256) T::AccountId
+			map hasher(blake2_128_concat) T::AccountId
```

Comment:
why is this one `blake2_128_concat`? I see that the user can set the controller, but logically the user can also decide the origin which dictates the key for `Bonded`, so what is the difference?

Reply:
(controller is actually indirectly mapped to accountId through StaticLookup path, maybe that makes an impact. )

Reply:
Because we know a stash account must call the functions to set the controller account, we know that the caller has a corresponding private key, and this imposes a restriction on the list of available addresses that can be used. A user would not be able to generate both the public and private key for an account with any arbitrary bytes. Hence a stash account would be okay to hash with `twox_64_concat`

However, when a stash sets a controller, they specify the controller account without any restrictions. A stash account could literally use any 32 bytes as the target for the controller. As a result, this could be a possible attack vector for creating an imbalanced trie, and thus we need to use `blake2_128_concat`